geo.go

```diff
@@ -0,0 +1,130 @@
+package tuna
+
+import (
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const IP2CUrl = "https://ip2c.org/"
+
+type Location struct {
+	IP          string `json:"IP"`
+	CountryCode string `json:"CountryCode"`
+	Country     string `json:"Country"`
+	City        string `json:"City"`
+}
+
+var emptyLocation = Location{}
+
+func (l *Location) Empty() bool {
+	if l == nil {
+		return true
+	}
+	return *l == emptyLocation
+}
+
+func (l *Location) Match(location *Location) bool {
+	if len(location.IP) > 0 && location.IP == l.IP {
+		return true
+	}
+	if len(location.CountryCode) > 0 && location.CountryCode == l.CountryCode {
+		return true
+	}
+	return false
+}
+
+type IPFilter struct {
+	Allow    []Location `json:"Allow"`
+	Disallow []Location `json:"Disallow"`
+}
+
+func (f *IPFilter) Empty() bool {
+	if f == nil {
+		return true
+	}
+
+	for _, loc := range f.Allow {
+		if !loc.Empty() {
+			return false
+		}
+	}
+	for _, loc := range f.Disallow {
+		if !loc.Empty() {
+			return false
+		}
+	}
+	return true
+}
+
+func getLocationFromIP2C(ip string) (*Location, error) {
+	queryUrl := IP2CUrl + ip
+	client := http.Client{
+		Timeout: 10 * time.Second,
+	}
+	unknown := &Location{CountryCode: "UNKNOWN"}
+	resp, err := client.Get(queryUrl)
+	if err != nil {
+		return unknown, err
+	}
+
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return unknown, err
+	}
+	loc, err := parseIP2C(string(body))
+	if err != nil {
+		return unknown, err
+	}
+	loc.IP = ip
+	return loc, nil
+}
+
+func parseIP2C(body string) (*Location, error) {
+	if body[0:1] != "1" {
```

Comment:
1. You need to handle `len(body) == 0` to avoid panic
2. `body[0]` is enough instead of `body[0:1]`

Reply:
body[0] will return uint8 rather than string, got this hint from IDE

Reply:
OK